Show events from user communities on dashboard

darrenvong opened the pull request:
<!---
Please describe the pull request below.
If it closes an issue, make sure to write "closes #1234"
If there is an issue but it isn't completely closed, still refer to the issue number, eg. "part of #1234"
--->
Probably my first time getting my hands dirty on the backend a little bit on this project, so hopefully I haven't done anything too terrible :p it was too inefficient and annoying me to have to send one request per community just to find out which events are in the communities the user is in, so I extended the `ListMyEvents` rpc to include events from the user's communities by default as well.

**Desktop**
![Screenshot 2021-09-28 at 02 21 33](https://user-images.githubusercontent.com/13669362/135006811-65adad61-5218-46d0-861f-e2639e64b5e5.png)

**Mobile**
![Screenshot 2021-09-28 at 02 22 07](https://user-images.githubusercontent.com/13669362/135006834-677732c7-8a3b-4acd-9a35-46d22b561228.png)

Closes #1880

<!---
Checklists - you can remove one that is not applicable (ie. remove backend checklist if you only worked on the web frontend)
If you need help with any of these, please ask :)
--->
**Backend checklist**
- [x] Formatted my code by running `autoflake -r -i --remove-all-unused-imports src && isort . && black .` in `app/backend`
- [x] Added tests for any new code or added a regression test if fixing a bug
- [ ] All tests pass - I'm not sure why the migration tests are failing for me locally in Docker, but it's fine in GitLab??
- [x] Run the backend locally and it works
- [ ] Added migrations if there are any database changes, rebased onto `develop` if necessary for linear migration history

**Web frontend checklist**
- [x] Formatted my code with `yarn format && yarn lint --fix`
- [x] There are no warnings from `yarn lint`
- [x] There are no console warnings when running the app
- [x] Added any new components to storybook
- [x] Added tests where relevant
- [x] All tests pass
- [x] Clicked around my changes running locally and it works
- [x] Checked Desktop, Mobile and Tablet screen sizes


<!---
Remember to request review from couchers-org/web, couchers-org/@backend or an individual.
Once your code is approved, remember to merge it if you have write access
--->

darrenvong committed:
feat: implement reduced version of upcoming events section

darrenvong committed:
fix: invalid div nested under p warning

darrenvong committed:
Extend ListMyEvents to include events from user communities

darrenvong committed:
Update web to use ListMyEvents RPC to reduce requests

darrenvong committed:
comment fix in ListMyEventsReq

darrenvong committed:
Update/fix ListMyEvents test

darrenvong committed:
fix: clamp event time to 2 lines on small mobile screens

Otherwise, users can barely any details of the time without most of it
being cropped off

darrenvong committed:
test: my events section

darrenvong committed:
fix: format in backend events test

darrenvong commented on app/backend/src/couchers/servicers/events.py:
```diff
@@ -734,6 +736,21 @@ def ListMyEvents(self, request, context):
+                        .where(ClusterSubscription.user_id == context.user_id)
+                        .where(Cluster.is_official_cluster)
+                        .order_by(Node.id)
+                        .limit(100000)
```
I kinda copied this from `threads.py` where 100000 is an arbitrarily "large enough" (?) number/page size cap so it should be sufficient to include all communities the user is in one page, but not unbounded

kalvdans replied:
Whenever we convert this into a subquery, the list of clusters will not be extracted from the database and this arbitrary limit will go away. So it is fine for now :)

darrenvong committed:
fix: revert unnecessary change on helper text 2

darrenvong committed:
test: storybook for my events section

kalvdans commented on app/backend/src/tests/test_events.py:
```diff
@@ -1,15 +1,14 @@
 import grpc
 import pytest
 from google.protobuf import wrappers_pb2
 from psycopg2.extras import DateTimeTZRange
+from sqlalchemy.sql.expression import update
 
 from couchers import errors
 from couchers.db import session_scope
+from couchers.models import EventOccurrence
 from couchers.utils import Timestamp_from_datetime, now, to_aware_datetime
```
Are these header rearrangements caused by a new version of `isort`? We should be doing them all at once in a separate single PR then.

darrenvong replied:
Since it's not pinned to a version in `requirements.txt`, I guess it's possible that I installed a newer version than the one you/Aapeli have on your machine as it's been the first time I really touched the backend. Maybe we should pin that to a version too? 🤷

kalvdans replied:
It might be a PR that was merged which we did not run `isort` on. I've opened https://github.com/Couchers-org/couchers/pull/2030 to add a check to CI.

Fine by me to do this change in this PR.

kalvdans commented on app/proto/events.proto:
```diff
@@ -255,6 +255,8 @@ message ListMyEventsReq {
   bool subscribed = 2;
   bool attending = 3;
   bool organizing = 4;
+  // include events in the user's communities
```
Great with comments!

kalvdans commented on app/backend/src/tests/test_events.py:
```diff
@@ -1741,9 +1744,12 @@ def new_event(hours_from_now, online=True):
         res = api.ListMyEvents(events_pb2.ListMyEventsReq(organizing=True))
         assert [event.event_id for event in res.events] == [e3, e4]
 
+        res = api.ListMyEvents(events_pb2.ListMyEventsReq(my_communities=True))
+        assert [event.event_id for event in res.events] == [e1, e2, e3, e4, e5]
```
Great with tests!

kalvdans reviewed (dismissed):
Backend looks good! There's still some failing frontend CI tests though.

darrenvong commented:
> There's still some failing frontend CI tests though.

It's because of this - https://gitlab.com/gitlab-com/gl-infra/production/-/issues/5590 😠  apparently the adjacent `web-next` job built and the tests from that passed though so I think we might be ok overall!

lucaslcode commented on app/web/src/features/dashboard/MyEvents.tsx:
```diff
@@ -0,0 +1,86 @@
+                />
+              ))}
+          </HorizontalScroller>
+          {hasNextPage && (
+            <StyledLink
+              className={classes.allUpcomingEventsLink}
+              to={eventsRoute}
```
This doesn't quite make sense to me - if there's more than one page of events in communities the user has joined, show a link to all upcoming events?

I know it's in the design but since there is no "My Events" page at this stage, I think it's better to make it an ordinary load more button (perhaps inside the `HorizontalScroller`).

darrenvong replied:
Do you mean have the "load more" button but the results will be horizontally scrollable on all screen sizes rather than going onto the next row below forever?

lucaslcode replied:
From memory `HorizontalScroller` does nothing above mobile screen sizes, so load more would add more horizontally on mobile and below otherwise.

But any way that works really - the main thing is not linking to the all events page with a label that makes it seem like its a "your events" page.

lucaslcode reviewed (changes requested):
One thing

darrenvong committed:
feat: swap show all upcoming events link to see more events button

The button loads more events on page rather than taking user to the
all events page based on PR comment

lucaslcode commented on app/web/src/features/dashboard/MyEvents.tsx:
```diff
@@ -70,12 +69,9 @@ export default function MyEvents() {
-            >
-              {SHOW_ALL_UPCOMING_EVENTS}
-            </StyledLink>
+            <Button onClick={() => fetchNextPage()}>
+              {SEE_MORE_EVENTS_LABEL}
+            </Button>
```
This is a bit weird on mobile (clicking the button makes more appear in the horizontal scroller).
How do you feel about an intersection observer div at the end of the horizontal scroller (only visible at mobile size) which triggers the load more (and displays a spinner?) Like in `InfiniteMessageLoader`.

darrenvong committed:
feat: infinite scroll to load events on small mobile screens

lucaslcode commented on app/web/src/features/dashboard/MyEvents.test.tsx:
```diff
@@ -89,4 +95,46 @@ describe("My events", () => {
+      ).not.toBeInTheDocument();
+
+      // Simulates scrolling horizontally to the end
+      mockIsIntersecting(screen.getByRole("progressbar"), true);
```
nice

lucaslcode commented on app/web/src/features/dashboard/MyEvents.tsx:
```diff
@@ -57,18 +64,23 @@ export default function MyEvents() {
               classes.cardContainer,
               classes.upcomingEventContainer
             )}
+            fetchNext={isBelowSm ? fetchNextPage : undefined}
+            hasMore={hasNextPage}
+            isFetching={isFetching}
```
Wow, I totally forgot I put this functionality in :D

darrenvong replied:
Indeed, made my job very easy 😉
